### PR TITLE
Added PointNet model

### PR DESCRIPTION
**Pull request by hummat**
* PointNet model for classification and segmentation
* Extended data augmentation transforms

---
This change is [Reviewable](https://reviewable.io/reviews/intel-isl/open3d-ml/156)

### CODE REVIEW

File: ml3d/torch/models/pointnet.py [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+        if self.task == 'classification':
+            inference_result = {
+                'predict_labels':
+                    np.array([results[0].argmax().detach().cpu().numpy()]),
+                'predict_scores':
+                    torch.nn.functional.softmax(results[0],
+                                                dim=1).detach().cpu().numpy()
+            }
+        else:
+            inference_result = {
```
Comment (sanskar107) [inline ml3d/torch/models/pointnet.py:155]: This is probably incorrect for segmentation, as we have to cover all of the points. Here, only a subsampled version gets covered.

**Review by sanskar107 — CHANGES_REQUESTED** [review]
Thanks for your contribution.
The inference part needs some modification for segmentation. You can try using samplers defined in `ml3d/datasets/samplers/semseg_random.py` and `ml3d/datasets/samplers/semseg_spatially_regular.py`. Please refer to other models (RandLANet and KPConv) for `sampler` class usage. They have been updated in `dev` branch.

Also have you tried training this model? does it integrate well with our training pipeline?

**Comment by germanros1987** [comment]
@hummat any updates on the feedback @sanskar107 provided?

File: ml3d/torch/models/pointnet.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+
+    def forward(self, inputs):
+        return self.net(inputs['point'].to(self.device))
+
+    def get_loss(self, Loss, results, inputs, device):
+        cfg = self.cfg
+        labels = inputs['data']['labels']
+
+        # Todo: Is this valid for classification?
+        scores, labels = filter_valid_label(results[0], labels, cfg.num_classes,
```
[the remark this answers is not in the document]
Comment (hummat) [reply ml3d/torch/models/pointnet.py:70]: Hm, seemed to work for me when I tested it. The classification labels are also just integers, except that they are the same for all points in the point cloud.

File: ml3d/torch/models/pointnet.py [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+        return data
+
+    # Todo: Parameters not consistent for base class and usage
+    def transform(self, data, attr):
+        point = data['point'].copy()
+
+        # Sub/super-sample to get consistent point cloud size for batching
+        # Todo: How to run semantic segmentation inference on different point cloud sizes?
+        choice = np.random.choice(len(point), self.num_points, replace=True)
+        point = point[choice, :]
```
[the remark this answers is not in the document]
Comment (hummat) [reply ml3d/torch/models/pointnet.py:100]: Yes, works for both.

File: ml3d/torch/models/pointnet.py (hunk 1 above)
Comment (hummat) [reply ml3d/torch/models/pointnet.py:155]: Oh, really? I though it was standard to only return predictions for points present in the input, i.e. if the input is subsampled, the output should be as well. Otherwise there needs to be some hand-crafted solution like interpolation to extend predicted labels to uncovered points. Is this desirable?

**Comment by hummat** [comment]
Thanks for the feedback @sanskar107. I don't understand what you mean by:

> "The inference part needs some modification for segmentation"

In the model class, there is only data transformation, preprocessing and batching happening. The sampling happens in the semantic segmentation pipeline, right? (RandLANet and KPConv also don't seem to make use of the `sampler` class).

Training for both semantic segmentation and classification was successfully tested using the semantic segmentation pipeline of the `master` branch, as the `dev` branch pipeline seems to be broken (train loop is commented out).

**Comment by germanros1987** [comment]
@sanskar107 please take a look when possible.

File: ml3d/torch/models/pointnet.py (hunk 1 above)
Comment (sanskar107) [reply ml3d/torch/models/pointnet.py:155]: The inference pipeline is expected to return exact same shape as that of input. Here, input is not subsampled, but the output is.
If you look at `inference_preprocess` method, you call `self.transform(...)` which will subsample the input.

**Review by sanskar107 — CHANGES_REQUESTED** [review]

**Comment by sanskar107** [comment]
> Thanks for the feedback @sanskar107. I don't understand what you mean by:
> 
> > "The inference part needs some modification for segmentation"
> 
> In the model class, there is only data transformation, preprocessing and batching happening. The sampling happens in the semantic segmentation pipeline, right? (RandLANet and KPConv also don't seem to make use of the `sampler` class).
> 
> Training for both semantic segmentation and classification was successfully tested using the semantic segmentation pipeline of the `master` branch, as the `dev` branch pipeline seems to be broken (train loop is commented out).

Thanks for pointing it out, PR #169 will fix training loop. The simplest way would be to store projected indices in `transform` and reproject back the output. But we recommend using sampler class(refer to PR #169). Your current inference pipeline randomly samples `num_points` number of points, but ideally we want to crop `num_points` from same region (maintaining density), run multiple iterations and merge different outputs.

**Comment by germanros1987** [comment]
@sanskar107 this may be ready for a re-review :)

**Comment by sanskar107** [comment]
@hummat thanks for your contribution. Everything looks good except the inference issue. Currently you randomly subsample the pointcloud in `transform` method which is also called by `inference_preprocess`. The segmentation output will correspond to the number of points in subsampled point cloud whereas it should be the number of points in input pointcloud. A quick fix would be to save the subsampling indices and later interpolate output in `inference_end` method. We are releasing a new version in next couple of days, we will really appreciate a quick fix.

**Comment by hummat** [comment]
> @sanskar107 this may be ready for a re-review :)

Hi, no it's not, sorry.

I've started to look into the inference sampling and I'm beginning to doubt that the PointNet model makes sense in your general framework. I think it wasn't really conceived with large scene point cloud segmentation in mind, only small part segmentation and classification, which doesn't seem to be the focus of this project. I guess we can just close this PR.
